### PR TITLE
Fix CVE-2020-1747 and CVE-2020-14343

### LINKED ISSUE #11045: Potential security issue
Hey there!

I belong to an open source security research community, and a member (@oivrip) has found an issue, but doesn’t know the best way to disclose it.

If not a hassle, might you kindly add a `SECURITY.md` file with an email, or another contact method? GitHub [recommends](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository) this best practice to ensure security issues are responsibly disclosed, and it would serve as a simple instruction for security researchers in the future.

Thank you for your consideration, and I look forward to hearing from you!

(cc @huntr-helper)

### LINKED ISSUE #11405: Resolve `sphinx-autodoc-typehints>=1.15` doc errors
## 📚 Documentation CI

sphinx-autodoc-typehints is causing typehint failures in the latest release.
The PR that caused it: https://github.com/tox-dev/sphinx-autodoc-typehints/pull/204

Temporary solution:
Pinning it to `<v1.15` is working.

Example failure: https://github.com/PyTorchLightning/pytorch-lightning/runs/4760961863?check_suite_focus=true
Probably coming somewhere from https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/profiler/pytorch.py

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @rohitgr7 @carmocca @akihironitta

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #11045

### Does your PR introduce any breaking changes? If yes, please list them.

If users were relying on the capabilities of `UnsafeLoader` when loading the `hparams`, this will be a breaking change.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda

### CODE REVIEW

File: tests/loggers/test_csv.py [hunk 1]
```diff
@@ -95,7 +94,6 @@ def test_file_logger_log_hyperparams(tmpdir):
         "bool": True,
         "dict": {"a": {"b": "c"}},
         "list": [1, 2, 3],
-        "namespace": Namespace(foo=Namespace(bar="buzz")),
```
Comment (carmocca) [inline tests/loggers/test_csv.py:98]: This requires `unsafe_load`

File: tests/loggers/test_csv.py [hunk 2]
```diff
@@ -95,7 +94,6 @@ def test_file_logger_log_hyperparams(tmpdir):
...
         "layer": torch.nn.BatchNorm1d,
```
Comment (carmocca) [inline tests/loggers/test_csv.py:97]: This is allowed with `full_load` but not `safe_load`.

Security is at odds with functionality here.

File: tests/loggers/test_csv.py (hunk 1 above)
Comment (justusschock) [reply tests/loggers/test_csv.py:98]: so we always need to convert namespaces to dicts now?

Comment (carmocca) [reply tests/loggers/test_csv.py:98]: Yes: https://github.com/yaml/pyyaml/issues/482

File: CHANGELOG.md [hunk 3]
```diff
@@ -273,6 +273,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Fixed security vulnerabilities CVE-2020-1747 and CVE-2020-14343 caused by the `PyYAML` dependency ([#11099](https://github.com/PyTorchLightning/pytorch-lightning/pull/11099))
```
Comment (carmocca) [inline CHANGELOG.md:276]: Should this PR be part of a bug-fix release? It's compatibility-breaking since we lose a degree of functionality.

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (awaelchli) [reply CHANGELOG.md:276]: I checked, we convert Namespace to dict in the save_hparams_to_yaml function. However, it will affect users who have a Namespace within that namespace and want to load the yaml file.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (justusschock) [reply CHANGELOG.md:278]: I think I'd stick to a feature release with that.

Comment (carmocca) [reply CHANGELOG.md:278]: Sounds good

**Review by Borda — APPROVED** [review]
can we check that it is fully back compatible with past checkpoints?

**Comment by carmocca** [comment]
> can we check that it is fully back compatible with past checkpoints?

@Borda It isn't if the checkpoint yaml requires arbitrary code execution.

**Review by ananthsub — APPROVED** [review]
